triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -217,54 +222,77 @@ class Runner(
-                  if (usedCommandIds.contains(commandId)) {
-                    throw new RuntimeException(s"Duplicate command id: $commandId")
-                  }
-                  usedCommandIds += commandId
-                  val commandUUID = UUID.randomUUID
-                  commandIdMap += (commandUUID -> commandId)
-                  val commandsArg = Commands(
-                    ledgerId = client.ledgerId.unwrap,
-                    applicationId = applicationId.unwrap,
-                    commandId = commandUUID.toString,
-                    party = party,
-                    commands = commands
-                  )
-                  logger.debug(
-                    s"submitting command ID $commandId, commands ${commands.map(_.command.value)}")
-                  submit(SubmitRequest(commands = Some(commandsArg)))
+  @throws[RuntimeException]
+  private def handleCommands[Z](
+      commandId: String,
+      commands: Seq[Command],
+      submit: SubmitRequest => Z): Z = {
+    if (usedCommandIds.contains(commandId)) {
+      throw new RuntimeException(s"Duplicate command id: $commandId")
+    }
+    usedCommandIds += commandId
+    val commandUUID = UUID.randomUUID
+    commandIdMap += (commandUUID -> commandId)
+    val commandsArg = Commands(
+      ledgerId = client.ledgerId.unwrap,
+      applicationId = applicationId.unwrap,
+      commandId = commandUUID.toString,
+      party = party,
+      commands = commands
+    )
+    logger.debug(s"submitting command ID $commandId, commands ${commands.map(_.command.value)}")
+    submit(SubmitRequest(commands = Some(commandsArg)))
+  }
+
+  import Runner.DamlFun
+
+  // Handles the value of update.
+  private def handleStepFreeResult(
+      clientTime: Timestamp,
+      v: SValue,
+      submit: SubmitRequest => Unit): SValue = {
+    def evaluate(se: SExpr) = {
+      val machine: Speedy.Machine =
+        Speedy.Machine.fromPureSExpr(compiledPackages, se)
+      // Evaluate it.
+      machine.setExpressionToEvaluate(se)
+      Machine.stepToValue(machine)
+    }
+    @tailrec def go(v: SValue): SValue = {
+      val resumed = unrollFree(v) match {
+        case Right(Right(vvv @ (variant, vv))) =>
+          vvv.match2 {
+            case "GetTime" /*(Time -> a)*/ => {
+              case DamlFun(timeA) => Right(evaluate(makeAppD(timeA, STimestamp(clientTime))))
+            }
+            case "Submit" /*(Commands, () -> a)*/ => {
+              case DamlTuple2(commands, DamlFun(unitA)) =>
+                converter.toCommands(commands) match {
+                  case Left(err) => throw new ConverterException(err)
+                  case Right((commandId, commands)) =>
+                    handleCommands(commandId, commands, submit)
                 }
-              }
+                Right(evaluate(makeAppD(unitA, SUnit)))
             }
-          case _ => {}
-        }
-        newState
+            case _ => {
+              case _ =>
```

Comment:
Why do we need the inner pattern match here?

Reply:
Just an artifact; we needed it when we returned a value, because that's the signature, but I suppose we can just

* [ ] throw earlier now.

Reply:
5709bffaf92efd91a1b92d751435959c4e86fc6d